### PR TITLE
SAT: testcase for checking if defined cursor is contained in its schema.

### LINKED ISSUE #3752: SAT should test a Stream's defined cursor is contained in its schema.
## Tell us about the problem you're trying to solve
I recently fixed a bug with the Stripe api where the defined cursor was not in the schema. This was causing incremental syncs to fail.

## Describe the solution you’d like
We should have tests that catch this, either as part of SAT or something else.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200368060087111/1200402175768903) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by avida**
## What
Resolves #3752 

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py [hunk 1]
```diff
@@ -114,6 +114,21 @@ def test_discover(self, connector_config, docker_runner: ConnectorRunner):
...
+        """
+        output = docker_runner.call_discover(config=connector_config)
+        catalog = filter_output(output, Type.CATALOG)[0].catalog
+        for stream in catalog.streams:
+            if stream.default_cursor_field:
+                schema = stream.json_schema
+                assert "properties" in schema, "Top level item should have an 'object' type for {stream.name} stream schema"
+                properties = schema["properties"].keys()
+                assert all(
+                    [cursor in properties for cursor in stream.default_cursor_field]
```
Comment (htrueman) [inline airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:129]: This won't work with deep level cursor fields.
- Assume we have `cursor_field=["f1", "f2"]`, which is the cursor path specified as a list of path parts.
- So this cursor field may be accessible with `schema["properties"]["fa1"]["f2"]`

File: airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+
+@pytest.mark.parametrize(
+    "schema, cursors, is_good",
+    [
+        ({}, ["created"], False),
+        ({"properties": {"created": {"type": "string"}}}, ["created"], True),
+        ({"properties": {"created_at": {"type": "string"}}}, ["created"], False),
+        ({"properties": {"created": {"type": "string"}}}, ["updated", "created"], False),
+        ({"properties": {"created": {"type": "string"}, "updated": {"type": "string"}}}, ["updated", "created"], True),
```
Comment (htrueman) [inline airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py:38]: Again, default_cursor_field may represent the cursor path, not 2 different values.

**Review by htrueman — CHANGES_REQUESTED** [review]
Multiple value `default_cursor_field` should be treated differently.
See https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py#L135-L137
Also `test_defined_cursors_exist_in_schema` may be simplified.

File: airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py (hunk 1 above)
Comment (avida) [reply airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py]: Got it, updated PR.

File: airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py (hunk 2 above)
Comment (avida) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py]: Fixed

**Review by htrueman — APPROVED** [review]

File: airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# SOFTWARE.
+#
+
+import pytest
+from airbyte_cdk.models import AirbyteStream
+from source_acceptance_test.tests.test_core import TestDiscovery as _TestDiscovery
+
+
+@pytest.mark.parametrize(
+    "schema, cursors, is_good",
```
Comment (sherifnada) [inline airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py:31]: rename `is_good` to something more descriptive :) maybe `should_fail`

**Review by sherifnada — APPROVED** [review]
🎉 

Please update the PR title to something descriptive that any engineer on the team can understand. See the documentation guide for more details.

File: airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py (hunk 3 above)
Comment (avida) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py]: Done

**Pull request merged**